### ISSUE TITLE
How to use theme object with typescript and @emotion/styled

### ISSUE THREAD

**Issue by codenakama** [chakra-ui/chakra-ui#90]
At the moment doing something like below throws an error `Property colors does not exist in Object`
```
import styled from '@emotion/styled'

const Div = styled.div`
    background: ${({theme}) => theme.colors.brand}
`
```

I've tried many approaches and nothing works but I've seen in other issues the library supports Typescript. Is there anything I am missing or doing incorrectly?

Cheers and thank your such an amazing piece of work.

**Comment by segunadebayo** [comment]
Hi @codenakama,

I'm not sure why you need to this. Chakra uses `styled-system` to make this easier.

Here's what you need to do:
```jsx
// Import Box
import { Box } from "@chakra-ui/core"

// Use Box
<Box bg="brand"/>
```

Behind the scenes, the `bg="brand"` prop is translated to `backgroundColor={theme.colors.brand}`

Use the `Box` component for most of your layout needs.

**Comment by lee-reinhardt** [comment]
In short, use `createStyled()` ([docs](https://emotion.sh/docs/typescript#define-a-theme)).

Example:
https://codesandbox.io/embed/tender-platform-f74f1

Although, the types for `ITheme` don't seem to be being inferred correctly beyond the first depth level. It knows all the root keys, like `colors`, and `shadow`, etc. But it thinks the values of all those are arrays, rather than objects, even though styled-system types look like they should be cool with an array or any generic object [definitely-typed source](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/styled-system/index.d.ts#L223).

<img width="828" alt="Screen Shot 2019-09-16 at 10 14 35 AM" src="https://user-images.githubusercontent.com/980089/64970153-db5dc780-d86a-11e9-8991-731263660576.png">

**Comment by codenakama** [comment]
@segunadebayo sorry bad example. for example let's say I want to handle a `selected` or `active` state or any other control variable to change css values, for example animating something if a prop is true.

**Comment by segunadebayo** [comment]
@codenakama, can you share a Codesandbox link of this scenario so I can help.

**Comment by codenakama** [comment]
@segunadebayo I will try @lee-reinhardt approach in a bit, to do codesandbox with full example I'd need more time, which I can do this evening but not now. cheers for the help though

**Comment by codenakama** [comment]
so just to update @lee-reinhardt suggestion worked but I think it's valuable to do the codesandbox. sorry I haven't got around doing this yet but I will do sometime this week. bear with me.

**Comment by segunadebayo** [comment]
Thanks. Feel free to share the codesandbox whenever you can.

**Comment by codenakama** [comment]
@segunadebayo I hope this helps, sorry it took a bit longer. I can consume the theme object but because it's not typed, typescript throws an err - in the codesandbox it's fine, but in my  project it actually won't compile if there is an error. 
Can you see anything I am doing wrong or need to config?
 - https://codesandbox.io/s/late-bird-9hzzp?fontsize=14

**Comment by xtalx** [comment]
@codenakama You can use the PseudoBox component for that to avoid all that extra work.

[forked sandbox with changes](https://codesandbox.io/s/divine-glade-69zpf)

**Comment by codenakama** [comment]
@xtalx that is true but, it does not actually address the reason why I opened this issue, which is no typing of the theme object when consuming it from styled component. The example I did there was with hover but imagine for handling selected/active states? 
I already have workarounds but isn't this a typing issue regardless?

**Comment by codenakama** [comment]
also if I can help solve this let me know. I am happy to scratch my itch and help fix it, I just don't know where to start.

**Comment by xtalx** [comment]
It does look like the types for theme may need to be tweaked as @lee-reinhardt mentioned above. That being said, I don't see a strong reason to ever need to use `styled` with this library. I don't think you'd ever need to pass the theme object around for styling individual properties or components even.

### TL;DR

Use the `css` function from [https://styled-system.com/css](@styled-system/css)
_Chakra should probably export this function from useTheme_

### Long answer

The PseudoBox covers all of the following emotion variants of css pseudo classes and attribute selectors.

| Selector                               | Prop           |
| -------------------------------------- | -------------- |
| `&:hover`                              | `_hover`       |
| `&:active`                             | `_active`      |
| `&:focus`                              | `_focus`       |
| `&:before`                             | `_before`      |
| `&:after`                              | `_after`       |
| `&::placeholder`                       | `_placeholder` |
| `&:first-of-type`                      | `_first`       |
| `&:last-of-type`                       | `_last`        |
| `[role=group]:hover &`                 | `_groupHover`  |
| `&:disabled`, `[aria-disabled=true]`   | `_disabled`    |
| `&[readonly]`, `&[aria-readonly=true]` | `_readonly`    |
| `&[aria-checked=true]`                 | `_checked`     |
| `&[aria-selected=true]`                | `_selected`    |
| `&[aria-expanded=true]`                | `_expanded`    |
| `&[aria-invalid=true]`                 | `_invalid`     |
| `&[aria-pressed=true]`                 | `_pressed`     |
| `&[aria-invalid=true]`                 | `_invalid`     |
| `&[aria-grabbed=true]`                 | `_grabbed`     |

You asked about responding to props and/or state, you can use props, state as expressions in the props of box. This shows all of the variations.
[sandbox](https://codesandbox.io/s/infallible-bhaskara-9o8er)

```jsx
/** @jsx jsx */
import React, { useState } from 'react'
import { jsx } from '@emotion/core'
import styled from '@emotion/styled'
import { Stack, Box, Button, PseudoBox, useTheme } from '@chakra-ui/core'
// @ts-ignore
import css from '@styled-system/css'

//css from styled system is probably the easiest
const Boop = styled(Box)(
  css({
    bg: 'green.500',
    '&:hover': {
      bg: 'red.500',
    },
  }),
)

export default function App() {
  const [active, setActive] = useState(false)
  const { colors } = useTheme()
  // Barring the fact that the theme types need to be tweaked
  // @ts-ignore
  const red = colors.red
  return (
    <React.Fragment>
      {/* CSS prop with styled system css function */}
      <h1 css={css({ fontSize: '2xl', '&:hover': { color: 'red.500' } })}>
        State is {active ? 'Active' : 'Inactive'}
      </h1>
      <Stack isInline>
        <PseudoBox
          height={200}
          width={200}
          // You can use state and props in Box style props
          bg={active ? 'green.500' : 'blue.500'}
          _hover={{ bg: active ? 'pink.500' : 'pink.900' }}
        >
          PseudoBox
        </PseudoBox>
        <Box
          height={200}
          width={200}
          // You can use state and props in Box style props
          bg={active ? 'blue.500' : 'green.500'}
          // using the css prop and values from them made accessible from useTheme()
          css={{ '&:hover': { background: active ? red['500'] : red['700'] } }}
        >
          css prop
        </Box>
        <Boop height={200} width={200}>
          Boop
        </Boop>
      </Stack>
      <Button onClick={() => setActive(!active)}>toggle active</Button>
    </React.Fragment>
  )
}

```

**Comment by segunadebayo** [comment]
Hi @codenakama,

Yes, we'll need to check the types for the theme. At the moment, we're extending the types from `@types/styled-system`.

For your use case, I'd suggest you try the following:

```jsx
// To style hover. This styled `&:hover`
<PseudoBox _hover={{bg: "red.400"}} />

// To style active. This styles `&:active`
<PseudoBox _active={{bg: "red.500"}} />

// To style selected. Use the `aria-selected` prop and you can style it using `_selected`
<PseudoBox aria-selected="true" _selected={{bg: "red.300"}} />

// To style disabled. Use the `aria-disabled` prop and you can style it using `_disabled`
<PseudoBox aria-disabled="true" _selected={{bg: "red.300"}} />
```

This way, you won't need to tackle the types issue with the theme.

Hope this helps.

**Comment by codenakama** [comment]
hey everyone, thanks for the feedback. Did learn a lot and just shows that the library is quite flexible and well thought from the ground. I will use this to cleanup my code and have it in a more declarative fashion.

**Comment by segunadebayo** [comment]
I'll close out this issue and a separate one for the theme type definition later.

**Comment by akiarostami** [comment]
Could you help me and tell me how I could change _before attributes on _hover? 

I can set _before and _hover separately, but I can't find a way to combine them. **_hover={{ _before: { ... }}}** didn't work.

**Comment by asterikx** [comment]
Hi, I want to add custom variants/styles and have a fully-typed theme. How can I use an extended chakra-ui theme with TypeScript? I can't find a follow-up issue on this @segunadebayo.

**Comment by erikzaadi** [comment]
@asterikx I haven't tried it yet, but `useTheme` get's a type definition so it should be possible to use as following:

```typescript
const { colors, myCustomStyle } = useTheme<MyCustomThemeType>()
```

**Comment by kimroen** [comment]
> @asterikx I haven't tried it yet, but `useTheme` get's a type definition so it should be possible to use as following:

`useTheme` doesn't accept a generic argument in the released version of the library, but I'm excited to hear it's coming ✨

**Comment by ofhouse** [comment]
What hinders you from creating the types the theme?

```ts
// theme.ts
import { extendTheme, withDefaultColorScheme } from '@chakra-ui/react';

const theme = extendTheme(
  {
    colors: {
      brand: {
        500: '#000000',
      },
    },
  },
  withDefaultColorScheme({ colorScheme: 'brand' })
);

type Theme = typeof theme;

export type { Theme };
export { theme };
```

```ts
// emotion.d.ts
import '@emotion/react';
import { Theme as CustomTheme } from './theme';

declare module '@emotion/react' {
  export interface Theme extends CustomTheme {}
}
``` 

Usage:
```tsx
import styled from '@emotion/styled'

const Div = styled.div`
    background: ${({theme}) => theme.colors.brand}
`
```
